Support experimental packages

conorsch opened the pull request:
First pass at adding support for the "test" branch of grsecurity patches, as described in #10. The implementation proposed in this PR works reasonably well, but does not prompt for vars authentication information for restricted "stable" downloads. In the interest of keeping the diffs readable for review, I'm submitting now, targeting `develop`, and will add variable prompts after further testing.

conorsch committed:
Renames "vars" directory to "defaults".

Most of the variables declared in the file won't
be overriden, therefore defaults/ is a more appropriate
location. The vars/ folder should contain platform-specific
variables, e.g. for certain OSes or stable/testing grsec patches.

conorsch committed:
Adds whitespace to handlebars variable interpolation.

Let the variables breathe!

conorsch committed:
Creates separate vars file for grsecurity test patches.

The defaults file now contains only the Linux tarball/sig URLs,
as well as the iterable list of files to download. The test
patches for grsecurity are in a separate file in vars/.
The flag and task to include the vars file has not been added yet.

conorsch committed:
Defaults to grsecurity test patches.

Adds default variable `grsecurity_patch_type` with value "test".
The URLs to the most recent test patches are imported from
a non-default vars file. Support for stable patches currently missing.

conorsch committed:
Uses OS detection to identify package list.

Supports Debian and Ubuntu, with slightly different
apt lists for both. Really, the apt list for Ubuntu
is specific to 14.04 LTS, so maybe it'd be wise to do
additional detection.

The role will fail if compilation is attempted on a
distro other than Debian or Ubuntu, with a helpful
message describing the problem and soliciting a PR. =)

conorsch committed:
Adds stable patch and sig URLs to separate vars file.

No logic yet for collecting credentials for download.

conorsch committed:
Creates separate task lists for downloading files.

Since there's conditional logic for both the grsecurity
patch type and the target distro (Debian/Ubuntu), it makes
sense to break out the logic into discrete task files,
and reference the condition on the include, same as
we do for vars files.

conorsch committed:
Bumps Linux kernel version 3.14.51 -> 3.14.52.

conorsch committed:
More verbose variable name for Linux kernel version.

Also adds whitespace on handlebars variable interpolation.

ageis commented on ansible/roles/build-grsec-kernel/tasks/fetch_files.yml:
```diff
@@ -1,50 +1,24 @@
-     cp {{download_directory}}/ubuntu-trusty/debian/control-scripts/p* \
-        {{download_directory}}/ubuntu-package/pkg/image/
-  when: target_os == 'ubuntu'
+- name: Fetch grsecurity patch and signature.
```
Can you explain why this is a nop?

conorsch replied:
Placeholder during development! The actual logic has moved to `tasks/fetch_grsecurity_files.yml`. Cleaned up the task lists in 989b7be. Thanks for catching!

ageis commented:
The changes in this PR look good to me. But for the test patch, don't you need to include downloads for the linux-4.1.7 kernel?

conorsch committed:
Renames separate task list for fetching Linux source.

Maintaining separate task lists for fetching:

 * Linux kernel source tarball and signature
 * grsecurity patch and signature (conditionally test/stable)
 * Ubuntu overlay for control files (conditional)

Having separate task lists makes the control flow in main.yml
more readable, and also helps expose discrete parts of the role
for use with conditional logic.

This commit also removes debugging messages left in during
development.

conorsch committed:
Use empty string as default value for grsecurity auth.

If left undefined, playbook will fail with an undefined warning.
When downloading the test patches, no auth is necessary, so
setting empty strings as auth values allows some `get_url` task
to be reused for both patching strategies.

conorsch committed:
Fetch Linux kernel source v4.x for grsecurity test patches.

The "test" patches for grsecurity apply only to the 4.x
kernels, so set the appropriate kernel version in the
grsecurity vars files.

conorsch committed:
Creates concatenated var for compilation directory.

conorsch commented:
@ageis You're right, the problem with the experimental patches was that the Linux kernel version wasn't being set appropriately. I've added logic to associate the "longterm" kernel version with the "stable" grsecurity patches, and the "stable" (`4.x`) kernel version with the "test" grsecurity patches.

Choosing kernel versions dynamically makes it difficult to hardcode the checksums for downloaded files, so I'll add additional tasks to handle fetching the remote file in future commits (perhaps better suited to a separate PR).

ageis commented:
@conorsch Excellent! With those two issues resolved, looks good to go now.

conorsch commented:
@ageis :metal:
